datafusion/expr/src/signature.rs

```diff
@@ -92,14 +92,22 @@ pub enum TypeSignature {
     /// A function such as `concat` is `Variadic(vec![DataType::Utf8, DataType::LargeUtf8])`
     Variadic(Vec<DataType>),
     /// One or more arguments of an arbitrary but equal type.
-    /// DataFusion attempts to coerce all argument types to match the first argument's type
+    /// DataFusion attempts to coerce all argument types to match to the common type with comparision coercion.
     ///
     /// # Examples
     /// Given types in signature should be coercible to the same final type.
     /// A function such as `make_array` is `VariadicEqual`.
     ///
     /// `make_array(i32, i64) -> make_array(i64, i64)`
     VariadicEqual,
+    /// One or more arguments of an arbitrary but equal type or Null.
+    /// Non-comparison coercion is attempted to match the signatures.
+    ///
+    /// Functions like `coalesce` is `VariadicEqual`.
```

Comment:
I am a little confused about what "Non-comparison coercion" means in this situation. Specifically how comparison coercion and non comparision coercion differ 🤔 

Does non comparison coercion mean "type union resolution" (aka `type_union_resolution`)? 

Also, I think `coalesce` is now   `VariadicEqualOrNull` (this says the opposite)

```suggestion
    /// One or more arguments of an arbitrary but equal type or Null.
    /// Non-comparison coercion is attempted to match the signatures.
    ///
    /// Functions like `coalesce` is `VariadicEqualOrNull`.
```

Reply:
yes. `Non-comparison coercion` is `type_union_resolution`

Actually, I think `type_union_resolution` is what we really need for coercion. Current `comparison_coercion` may not exist in the long term because I think there are many `workaround` solutions included in it, and those are possible to pull out to their **correct places**.

Ideally, we should shape `comparison_coercion` to more like `type_union_resolution`. But, adjusting logic inside it is vulnerable if the test is not covered.


At least I can tell the dict-coercion is different between these two now.